### PR TITLE
Feature/improve nav generation script and add outline file

### PR DESCRIPTION
**Pull request by Mathieudewet**
Update doc generation scripts.

### CODE REVIEW

File: bin/generate-nav [hunk 1]
```diff
@@ -1,52 +1,62 @@
...
-      title,
-      anchors: normalizeAnchors(anchors),
-    };
+const lvl =  '  ';
+const generatedItemsId = `${lvl.repeat(2)}- id: `;
+const generatedItemsTitle = `${lvl.repeat(3)}title: `;
+const generatedAnchorId = `${lvl.repeat(4)}- id: `;
+const generatedAnchorTitle = `${lvl.repeat(5)}title: `;
+
+let yaml = '';
```
Comment (dunglas) [inline bin/generate-nav]: Create an object, that you'll dump in YAML at the end using `safeDump`. The code will be easier to read, and you'll avoid a lot of issues including dealing with special characters...

File: bin/generate-nav [hunk 2]
```diff
@@ -1,52 +1,62 @@
...
+  if (node === 'h1') {
+    yaml += !isCurrentLvl ? `${generatedItemsTitle}"${data}"\n` : `${generatedItemsTitle}"${data}"\n`;
+  } else {
+    yaml += isChild ? `${lvl.repeat((parseInt(node.charAt(1)) * 2) - 1)}anchors: \n${lvl.repeat(((parseInt(node.charAt(1) - 2) * 2)))}${generatedAnchorId}${(data.toLowerCase()).replace(/[^A-Za-z0-9\-\ \_]/g, '').replace(/ /g, '-')}\n${lvl.repeat(((node.charAt(1) - 2) * 2))}${generatedAnchorTitle}${data}\n` : `${lvl.repeat(((node.charAt(1) - 2) * 2))}${generatedAnchorId}${(data.toLowerCase()).replace(/[^A-Za-z0-9\-\ \_]/g, '').replace(/ /g, '-')}\n${lvl.repeat(((node.charAt(1) - 2) * 2))}${generatedAnchorTitle}${data}\n`;
+  }
+}
+
+loadedOutline.map((section) => {
+  processDirectory(section.title, section.path);
+  var itemExist = false;
```
Comment (dunglas) [inline bin/generate-nav]: `let`

File: bin/generate-nav [hunk 3]
```diff
@@ -1,52 +1,62 @@
...
+
+  section.items.map( (item) => {
+    yaml += itemExist ? `${generatedItemsId}${item.id}\n` : `${lvl.repeat(1)}items: \n${generatedItemsId}${item.id}\n`;
+    itemExist = true;
+    var htmlLvl = 'h1';
+
+    Object.keys(item.yaml.childNodes).map(function(objectKey, index) {
```
Comment (dunglas) [inline bin/generate-nav]: You can use an arrow function here.

File: bin/generate-nav [hunk 4]
```diff
@@ -1,52 +1,62 @@
...
+      if (Math.abs(index % 2) == 0) {
```
Comment (dunglas) [inline bin/generate-nav]: What is this code doing? A comment?

File: bin/generate-outline [hunk 5]
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env node
+
+const htmlFromMdFile = require('./htmlFromMdFile');
+const { lstatSync, readdirSync, readFileSync, writeFile } = require('fs');
+const { resolve } = require('path');
+const jsYaml = require('js-yaml');
+const generatedTitle = '- title: ';
```
Comment (dunglas) [inline bin/generate-outline]: Same here, use `safeDump`

File: bin/generate-outline [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        h1 = h1.replace(/<\/h1>/, '"');
+        yaml += h1;
+        let h2 = `${htmlFromMdFile.generateHtmlFromMdFileNode(loadedOutline[key]['path'], id, 'h2')}`;
+
+        if (h2.length > 0) {
+            yaml += generatedItemsAnchors;
+        }
+        h2s = h2.split('<h2>');
+        h2s.forEach(function (el) {
+            if (el.length > 0) {
```
Comment (dunglas) [inline bin/generate-outline]: Invert the test to use a guard clause

File: bin/generate-outline [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+for(var key in loadedOutline) {
+    if (!lstatSync(`${__dirname}/../src/pages/docs/${loadedOutline[key]['path']}`).isDirectory()) {
+        continue;
+    }
+    let files = readdirSync(`${__dirname}/../src/pages/docs/${loadedOutline[key]['path']}`);
+
+    yaml += `${generatedTitle}'${loadedOutline[key]['title']}'\n`;
+    yaml += `${generatedPath}${loadedOutline[key]['path']}\n`;
+    yaml += generatedItems;
+    Object.keys(loadedOutline[key]['items']).forEach(function(itemEntry) {
```
Comment (dunglas) [inline bin/generate-outline]: You can use an arrow function

File: bin/generate-outline [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+const loadedOutline = jsYaml.safeLoad(readFileSync(`${__dirname}/../config/outline.yaml`, 'utf8'));
+
+let yaml = '';
+let h2s = [];
+
+for(var key in loadedOutline) {
+    if (!lstatSync(`${__dirname}/../src/pages/docs/${loadedOutline[key]['path']}`).isDirectory()) {
+        continue;
+    }
+    let files = readdirSync(`${__dirname}/../src/pages/docs/${loadedOutline[key]['path']}`);
```
Comment (dunglas) [inline bin/generate-outline]: `const`?

File: bin/generate-outline (hunk 6 above)
Comment (Mathieudewet) [reply bin/generate-outline]: Not sure i'm getting what you mean here, test opposite and continue if it's the case ?

I mean like 

```
if (! el.length > 0) {
 continue;
}
el = el.replace(/\\/, '\\\\');
...
```

File: bin/generate-nav [hunk 9]
```diff
@@ -1,52 +1,62 @@
...
-        title,
-      };
-    }
-
-    return {
-      id,
-      title,
-      anchors: normalizeAnchors(anchors),
-    };
+const lvl =  '  ';
```
[the remark this answers is not in the document]
Comment (Mathieudewet) [reply bin/generate-nav]: I had some trouble to handle it with safeDump that's why i did like so, maybe i misused it.

File: bin/generate-nav [hunk 10]
```diff
@@ -1,52 +1,62 @@
...
+    Object.keys(elHtml).forEach(key => elHtml[key] === undefined ? delete elHtml[key] : '');
+    item.yaml = jsyaml.safeLoad(jsyaml.safeDump(elHtml));
   });
+});
+
+function processDirectory(title, path) {
+  yaml += `- title: "${title}"\n${lvl.repeat(1)}path: ${path}\n`;
+}
+
+function processNode(node, data = null, isCurrentLvl = false, isChild) {
```
[the remark this answers is not in the document]
Comment (Mathieudewet) [reply bin/generate-nav]: In fact I refactored it severals time and let it but it's (isCurrentLvl) not usefull anymore.

[the remark this answers is not in the document]
Comment (Mathieudewet) [reply bin/generate-nav]: Is Ischild confusing ? is IsNextLvl more meaningfull ?

File: bin/generate-outline (hunk 6 above)
Comment (dunglas) [reply bin/generate-outline]: yes

File: bin/fileProcessor [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    .replace(/```(|yaml|xml|php|json|javascript|sh)[^`]*```/gm, '');
+
+  return markdown.toHTML(content);
+}
+
+function getHTML5Outline(fileContent) {
+  return HTML5Outline(new JSDOM(fileContent).window.document.body);
+}
+
+function getTree(tree, node, id) {
```
Comment (toofff) [inline bin/fileProcessor]: this should work, to test ;)

```
function getTree(tree, node, id) {
  if (!node.sections || !tree) {
    return tree;
  }

  return [...tree, ...node.sections.map((section) => {
    const toPush = {
      id: 'H1' !== section.heading.tagName ? slug(section.heading.innerHTML) : id,
      title: section.heading.innerHTML,
    };

    return 0 < getTree([], section).length ? { ...toPush, anchors: getTree([], section) } : toPush;
  })];
}
```

File: bin/generate-nav [hunk 12]
```diff
@@ -1,52 +1,22 @@
...
-        title,
-      };
-    }
-
-    return {
-      id,
-      title,
-      anchors: normalizeAnchors(anchors),
-    };
+loadedOutline.map(directory => {
```
Comment (toofff) [inline bin/generate-nav]: `loadedOutline.forEach(directory => {`

File: bin/check-outline [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+const loadedOutline = jsYaml.safeLoad(readFileSync(`${__dirname}/../config/outline.yaml`, 'utf8'));
+const directories = (function () {
+  return readdirSync(docPagesDirectory)
+    .filter(folder => statSync(join(docPagesDirectory, folder))
+      .isDirectory() && '.' !== folder[0]);
+}());
+
+const outlineDirectories = Object.keys(loadedOutline).map(key => loadedOutline[key].path);
+const oddDirectories = {
+  missingDirectories: directories.filter(f => !outlineDirectories.includes(f)),
```
Comment (toofff) [inline bin/check-outline]: the name of the `f` variable is too short

File: bin/check-outline [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+  surplusDirectories: outlineDirectories.filter(f => !directories.includes(f)),
```
Comment (toofff) [inline bin/check-outline]: the name of the `f` variable is too short

File: bin/check-outline [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+};
+
+function updateOutline(oddDirectories) {
```
Comment (toofff) [inline bin/check-outline]: the method is useless

```
if (oddDirectories.missingDirectories.length > 0) {
  console.warn('\x1b[31m', `Following Directories are missing in Outline, please add them with relevants titles: ${oddDirectories.missingDirectories.join(', ')}`, '\x1b[37m');
}
const newOutline = loadedOutline.filter(folder => !oddDirectories.surplusDirectories.includes(folder.path));
```

File: bin/check-outline [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+  if (oddDirectories.missingDirectories.length > 0) {
+    console.warn('\x1b[31m', `Following Directories are missing in Outline, please add them with relevants titles: ${oddDirectories.missingDirectories.join(', ')}`, '\x1b[37m');
+  }
+  return loadedOutline.filter(folder => !oddDirectories.surplusDirectories.includes(folder.path));
+}
+
+const newOutline = updateOutline(oddDirectories);
```
Comment (toofff) [inline bin/check-outline]: delete please

File: bin/check-outline [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+newOutline.forEach(function (item) {
```
Comment (toofff) [inline bin/check-outline]:
```
newOutline
  .filter((item) => {
    const path = `${__dirname}/../src/pages/docs/${item.path}`;

    return lstatSync(path).isDirectory();
  })
  .forEach((item) => {
    const files = readdirSync(`${__dirname}/../src/pages/docs/${item.path}`).filter(file => file.endsWith('.md'));

    Object.keys(item.items)
      .filter((itemEntry) => !files.includes(`${Object.values(item.items[itemEntry])}.md`))
      .forEach((itemEntry) => {
        console.warn('\x1b[31m', `missing ${Object.values(item.items[itemEntry])}.md file in src/pages/docs/${item.path}`, '\x1b[37m');
      });

    const newOutlineFiles = item.items.map(file => `${file.id}.md`)

    files
      .filter((file) => !newOutlineFiles.includes(file))
      .forEach((file) => {
        console.warn('\x1b[31m', `missing file ${file} in ${item.path} section in config/outline.yaml`, '\x1b[37m');
      });
  });
```

File: bin/check-outline [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+  const newOutlineFiles = item.items.map(file => `${file.id}.md`)
+
+  files.forEach(function (file) {
+    if (!newOutlineFiles.includes(file)) {
+      console.warn('\x1b[31m', `missing file ${file} in ${item.path} section in config/outline.yaml`, '\x1b[37m');
+    }
+  });
+});
+
+writeFile(resolve(__dirname, '../config/outline.yaml'), jsYaml.safeDump(newOutline), error => error && console.warn(error));
```
Comment (toofff) [inline bin/check-outline]:
```
writeFile(
  resolve(__dirname, '../config/outline.yaml'),
  jsYaml.safeDump(newOutline),
  (error) => error && console.warn(error)
);
```

File: bin/fileProcessor [hunk 19]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env node
+
```
Comment (toofff) [inline bin/fileProcessor]: Create helper file please.

File: bin/generate-nav [hunk 20]
```diff
@@ -1,52 +1,22 @@
...
-let content = fs.readFileSync(path.resolve(__dirname, '../src/pages/docs/index.md')).toString();
-content = content.replace(/`/gm, '"');
-content = markdown.toHTML(content);
+const loadedOutline = jsyaml.safeLoad(readFileSync(`${__dirname}/../config/outline.yaml`, 'utf8'));
 
-const html = new DOMParser().parseFromString(content);
+let yaml = '';
 
-const guessPath = h2 => xpath.select('following-sibling::ol[1]/li[1]/a/@href', h2)[0].value.split('/')[0];
+loadedOutline.forEach((directory) => {
```
Comment (toofff) [inline bin/generate-nav]:
```
const yaml = loadedOutline
  .map((directory) => {
    const items = directory.items.map((file) => fileProcessor.processFile(directory.path, file.id)[0]);
    return jsyaml.safeDump([{ ...directory, items }], { lineWidth: 255 });
  })
  .join('');
```

File: bin/generate-nav [hunk 21]
```diff
@@ -1,52 +1,22 @@
...
-const normalizeTitle = title => title.trim().split(':')[0];
-
-const results = xpath.select('//h2', html).map(h2 => ({
-  title: normalizeTitle(h2.firstChild.data),
-  path: guessPath(h2),
-  items: normalizeItems(xpath.select('following-sibling::ol[1]/li', h2)),
-}));
-
-fs.writeFile(path.resolve(__dirname, '../src/pages/docs/nav.yml'), yaml.dump(results), error => error && console.warn(error));
+writeFile(
```
Comment (toofff) [inline bin/generate-nav]: `writeFile(resolve(__dirname, '../src/pages/docs/nav.yml'), yaml, (error) => error && console.warn(error));`

File: gatsby-node.js [hunk 22]
```diff
@@ -42,33 +48,80 @@ exports.createPages = ({ boundActionCreators, graphql }) => {
...
+      };
+    })];
+  }
+
   return Promise.all([navQuery, docQuery]).then((values) => {
     const nav = values[0].data.allNavYaml.edges;
     const docs = values[1].data.allMarkdownRemark.edges;
     const parseNav = [];
 
     nav.map((navItem) => {
```
Comment (toofff) [inline gatsby-node.js]:
```
nav
  .filter((navItem) => navItem.node.items)
  .forEach((navItem) => {
    const { path, title, items } = navItem.node;

    [...parseNav, items.map((subItem) => {
      const path = `docs/${path}/${subItem.id}`;

      return {
        path,
        title: subItem.title,
        rootPath: title,
        items: getNav([], path, subItem)
      };
    });
  });
```

File: gatsby-node.js [hunk 23]
```diff
@@ -94,17 +152,19 @@ exports.createPages = ({ boundActionCreators, graphql }) => {
...
-        },
+          html,
+          nav
+        }
       });
 
-      if(redirect) {
+      if (redirect) {
         const redirects = [`/${redirect}`, `/${redirect}/`, `/${path}/`];
         redirects.map(redirPath =>
```
Comment (toofff) [inline gatsby-node.js]: Use .forEach

File: gatsby-node.js [hunk 24]
```diff
@@ -83,9 +136,14 @@ exports.createPages = ({ boundActionCreators, graphql }) => {
...
         }
 
-        current = { path: current.path, title: `${current.rootPath} - ${current.title}` };
+        current = {
+          path: current.path,
+          title: `${current.rootPath} - ${current.title}`
+        };
       }
 
+      const nav = jsyaml.safeLoad(
```
Comment (toofff) [inline gatsby-node.js]: `const nav = jsyaml.safeLoad(readFileSync(`${__dirname}/src/pages/docs/nav.yml`, 'utf8'));`

File: gatsby-node.js [hunk 25]
```diff
@@ -42,33 +48,80 @@ exports.createPages = ({ boundActionCreators, graphql }) => {
...
+        parseNav.push({
+          path: currentPath,
+          title: subItem.title,
+          rootPath: title,
+          items: getNav([], currentPath, subItem)
+        });
+      });
     });
 
     docs.map((edge) => {
```
Comment (toofff) [inline gatsby-node.js]: Use .forEach

File: gatsby-node.js [hunk 26]
```diff
@@ -42,33 +48,80 @@ exports.createPages = ({ boundActionCreators, graphql }) => {
...
+
+      const regex = RegExp(/<a href="#((?:[\w-]+(?:-[\w-]+)*)+)" aria-hidden="true" class="anchor">/, 'gm');
+      let tmpResult;
+      let result = [];
+
+      while ((tmpResult = regex.exec(html)) !== null) {
+        result.push(tmpResult[1]);
+      }
+
+      headings.forEach(function (currentVal, index) {
```
Comment (toofff) [inline gatsby-node.js]: `headings.forEach((currentVal, index) => {`

File: gatsby-node.js [hunk 27]
```diff
@@ -42,33 +48,80 @@ exports.createPages = ({ boundActionCreators, graphql }) => {
...
+          items: getNav([], currentPath, subItem)
+        });
+      });
     });
 
     docs.map((edge) => {
       const path = edge.node.fields.path;
       const redirect = edge.node.fields.redirect;
-      const index = parseNav.findIndex(element => element.path === path || element.path === redirect);
+      const index = parseNav.findIndex(
```
Comment (toofff) [inline gatsby-node.js]: `const index = parseNav.findIndex((element) => element.path === path || element.path === redirect);`

File: gatsby-node.js [hunk 28]
```diff
@@ -117,13 +177,12 @@ exports.onCreateNode = ({ node, boundActionCreators, getNode }) => {
...
   }
-
   const fileNode = getNode(node.parent);
   let nodePath = fileNode.relativePath.replace('.md', '');
   let html = node.internal.content;
-  const re = /(\]\((?!http)(?!#)(.*?)\))/gi;
-  const localUrls = [];
+  let localUrls = [];
   let matches;
+  const re = /(\]\((?!http)(?!#)(.*?)\))/gi;
```
Comment (toofff) [inline gatsby-node.js]: the name of the `re` variable is too short

File: bin/check-outline [hunk 29]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env node
+
+const { docPagesDirectory } = require('../constants');
+const { lstatSync, readdirSync, readFileSync, writeFile, statSync } = require('fs');
+const { join, resolve } = require('path');
+const jsYaml = require('js-yaml');
+
+const loadedOutline = jsYaml.safeLoad(readFileSync(`${__dirname}/../config/outline.yaml`, 'utf8'));
+const directories = (function () {
```
[the remark this answers is not in the document]
Comment (Mathieudewet) [reply bin/check-outline]: It doesnt seem to work as it.

[the remark this answers is not in the document]
Comment (dunglas) [reply bin/check-outline]: `const directories = (() => readdirSync(docPagesDirectory)
    .filter((f) => statSync(join(docPagesDirectory, f)).isDirectory() && '.' !== f[0]);)()`

File: bin/fileProcessor (hunk 19 above)
Comment (toofff) [reply bin/fileProcessor]: @Mathieudewet ping

File: README.md [hunk 30]
```diff
@@ -18,14 +18,14 @@ Be sure to install them first and to have the binaries in your path.
...
-
-bin/retrieve-documentation
-
 # Install dependencies
 
 yarn install
 
+# Retrieve documentation
+
+bin/retrieve-documentation
```
Comment (toofff) [inline README.md]: Rebase please. This change has been made in another pull request

Comment (toofff) [reply README.md]: See https://github.com/api-platform/website/pull/43/files

**Pull request merged**